cabal-install/Distribution/Client/Dependency.hs

```diff
@@ -361,7 +361,7 @@ dontUpgradeNonUpgradeablePackages params =
   where
     extraConstraints =
       [ LabeledPackageConstraint
-        (PackageConstraint (scopeToplevel pkgname) PackagePropertyInstalled)
+        (PackageConstraint (ScopeAnyQualifier pkgname) PackagePropertyInstalled)
```

Comment:
`scopeTopLevel` was only added in #4219 (originally `unqualified`).  Before that, every constraint applied to all occurrences of the package, so the constraints above always prevented `base`, etc. from being installed.